### PR TITLE
Along over many 5

### PR DESCRIPTION
**Pull request by hgrano**
Adds functionality to use `Along` and `Over` with multiple dimensions e.g. `Along(_0, _1)`.
`Position`s can also be indexed at multiple dimensions e.g. `Position("hello", 1, true)(_0 :: _1 :: HNil)` --> `StringValue("hello") :: IntValue(1) :: HNil`.
Multiple dimensions can be removed from `Position`s in a single call e.g. `Position("hello", 1, true).remove(_0 :: _1 :: HNil)` --> `Position(BooleanValue(true))`
Old API is left almost unchanged. Old `Along`, `Over` and index/remove calls can be left as is.

### CODE REVIEW

File: grimlock-core/src/main/scala/commbank/grimlock/framework/Slice.scala [hunk 1]
```diff
@@ -28,71 +28,184 @@ trait Slice[P <: HList, S <: HList, R <: HList] {
...
+  dimensions: I
 )(implicit
-  ev1: Position.IndexConstraints.Aux[P, D, V],
-  ev2: Position.RemoveConstraints.Aux[P, D, Q]
-) extends Slice[P, V :: HNil, Q] {
-  def selected(pos: Position[P]): Position[V :: HNil] = Position(pos(dimension))
-  def remainder(pos: Position[P]): Position[Q] = pos.remove(dimension)
+  ev1: Position.IndexConstraints.Aux[P, I, R],
+  ev2: Position.RemoveConstraints.Aux[P, I, Q],
+  ev3: Position.ValueConstraints[P]
```
Comment (poojke) [inline grimlock-core/src/main/scala/commbank/grimlock/framework/Slice.scala]: Question: is it possible for `ev3` to be removed (or moved to either `IndexConstraints` or `RemoveConstraints`)?

File: grimlock-core/src/main/scala/commbank/grimlock/framework/Slice.scala [hunk 2]
```diff
@@ -28,71 +28,184 @@ trait Slice[P <: HList, S <: HList, R <: HList] {
...
+    ev4: Witness.Aux[D]
+  ): Along[P, D :: HNil, ev1.V, ev2.Q] = Along(ev4.value :: HNil)(ev1, ev2, ev3)
+
+  /** Construct an `Along` for 2 dimensions using types. */
+  def apply[
+    P <: HList,
+    D <: Nat,
+    E <: Nat
+  ](implicit
+    ev1: Position.IndexConstraints[P, D :: E :: HNil] { type V <: HList},
```
Comment (poojke) [inline grimlock-core/src/main/scala/commbank/grimlock/framework/Slice.scala]: Minor: missing whitespace before closing `}`

File: grimlock-core/src/main/scala/commbank/grimlock/framework/Slice.scala [hunk 3]
```diff
@@ -28,71 +28,184 @@ trait Slice[P <: HList, S <: HList, R <: HList] {
...
  *
- * @param dimension Dimension of the selected coordinate.
+ * @param dimensions Dimensions of the selected coordinates.
  */
 case class Over[
   P <: HList,
-  D <: Nat,
-  V <: Value[_],
+  I <: HList,
+  R <: HList,
```
Comment (poojke) [inline grimlock-core/src/main/scala/commbank/grimlock/framework/Slice.scala]: Question: Should `R` and `Q` be renamed to `S` and `R` respectively? So they match their name/purpose in the `Slice` types?

File: grimlock-core/src/main/scala/commbank/grimlock/framework/Position.scala [hunk 4]
```diff
@@ -384,22 +388,59 @@ object Position {
...
+      ev2: Lazy[Aux[P, E :: I, OHT :: W]],
+      ev3: GT[E, D] // indices must be in ascending order
+    ): Aux[P, D :: E :: I, OH :: OHT :: W] = new IndexConstraints[P, D :: E :: I] {
+      type V = OH :: OHT :: W
 
-    /** Implicit with all constraints for indexing coordinates in a position. */
-    implicit def indexConstraints[
+      implicit val vTag: ClassTag[V] = implicitly[ClassTag[V]]
+
+      def index(coordinates: P, dimension: D :: E :: I): V = ev1.index(coordinates, dimension.head) :: ev2.value.index(coordinates, dimension.tail)
```
Comment (poojke) [inline grimlock-core/src/main/scala/commbank/grimlock/framework/Position.scala]: Minor: there are a few lines that are >120 chars

File: grimlock-core/src/main/scala/commbank/grimlock/framework/Position.scala [hunk 5]
```diff
@@ -600,20 +640,55 @@ object Position {
     def apply[P <: HList, V <: Value[_]](implicit ev: PrependConstraints[P, V]): Aux[P, V, ev.Q] = ev
   }
 
-  /** Type that captures all constraints for removing coordinates from a position. */
-  trait RemoveConstraints[P <: HList, D <: Nat] extends java.io.Serializable {
+  /** Type that captures all constraints for removing coordinate(s) from a position. */
+  trait RemoveConstraints[
```
Comment (poojke) [inline grimlock-core/src/main/scala/commbank/grimlock/framework/Position.scala]: Minor: this probably fits on a single line

File: grimlock-core/src/main/scala/commbank/grimlock/framework/Slice.scala (hunk 3 above)
Comment (hgrano) [reply grimlock-core/src/main/scala/commbank/grimlock/framework/Slice.scala]: Probably yes. Assuming we rename them, then we would have `Over[P, I, S, R]` and `Along[P, I, R, S]`. Is that ok? As in, the ordering of `S` and `R`

File: grimlock-core/src/main/scala/commbank/grimlock/framework/Slice.scala (hunk 1 above)
Comment (hgrano) [reply grimlock-core/src/main/scala/commbank/grimlock/framework/Slice.scala]: I've done that and now it works without `ev3`. There's now an `implicit val x: Position.ValueConstraints[P]` on the `IndexConstraints`. Technically I think we can remove that `implicit val` too. Might maybe you might want access that implicit for some reason?

File: grimlock-core/src/main/scala/commbank/grimlock/framework/Slice.scala (hunk 2 above)
Comment (hgrano) [reply grimlock-core/src/main/scala/commbank/grimlock/framework/Slice.scala]: Fixed

File: grimlock-core/src/main/scala/commbank/grimlock/framework/Position.scala (hunk 4 above)
Comment (hgrano) [reply grimlock-core/src/main/scala/commbank/grimlock/framework/Position.scala]: I've fixed the lines that I added which were over 120 chars

File: grimlock-core/src/main/scala/commbank/grimlock/framework/Position.scala (hunk 5 above)
Comment (hgrano) [reply grimlock-core/src/main/scala/commbank/grimlock/framework/Position.scala]: Done

File: grimlock-core/src/main/scala/commbank/grimlock/framework/Slice.scala (hunk 3 above)
Comment (poojke) [reply grimlock-core/src/main/scala/commbank/grimlock/framework/Slice.scala]: Seems reasonable.

File: grimlock-core/src/main/scala/commbank/grimlock/framework/Position.scala (hunk 4 above)
Comment (hgrano) [reply grimlock-core/src/main/scala/commbank/grimlock/framework/Position.scala]: Just check the formatting is still what you expect. I think it's ok

File: grimlock-core/src/main/scala/commbank/grimlock/framework/Slice.scala (hunk 1 above)
Comment (poojke) [reply grimlock-core/src/main/scala/commbank/grimlock/framework/Slice.scala]: See if you can remove it from the index constraints as well then. If it's not possible then just leave it as is.

File: grimlock-core/src/main/scala/commbank/grimlock/framework/Position.scala [hunk 6]
```diff
@@ -384,38 +388,84 @@ object Position {
...
+      ev2: Lazy[Aux[P, E :: I, OHT :: W]],
+      ev3: GT[E, D] // indices must be in ascending order
+    ): Aux[P, D :: E :: I, OH :: OHT :: W] = new IndexConstraints[P, D :: E :: I] {
+      type V = OH :: OHT :: W
+
+      implicit val vTag: ClassTag[V] = implicitly[ClassTag[V]]
+
+      implicit val valueConstraints: ValueConstraints[P] = ev1.valueConstraints
+
+      def index(coordinates: P, dimension: D :: E :: I): V =
```
Comment (poojke) [inline grimlock-core/src/main/scala/commbank/grimlock/framework/Position.scala]: Would prefer the following:
```
      def index(
        coordinates: P,
        dimension: D :: E :: I
      ): V = ev1.index(coordinates, dimension.head) :: ev2.value.index(coordinates, dimension.tail)
```

File: grimlock-core/src/main/scala/commbank/grimlock/framework/Position.scala [hunk 7]
```diff
@@ -600,20 +650,53 @@ object Position {
...
+      OUT <: HList,
+      I <: HList
+    ](implicit
+      ev1: Lazy[Aux[P, E :: I, R]], // remove the indices in the tail first
+      ev2: Aux[R, D, OUT], // use the instance for the single Nat case (`removeNat`)
+      ev3: GT[E, D] // indices should be in ascending order
+    ): Aux[P, D :: E :: I, OUT] = new RemoveConstraints[P, D :: E :: I] {
+      type Q = OUT
+
+      def remove(coordinates: P, dimension: D :: E :: I): Position[Q] =
```
Comment (poojke) [inline grimlock-core/src/main/scala/commbank/grimlock/framework/Position.scala]: Would prefer same formatting as in previous comment.

File: grimlock-core/src/main/scala/commbank/grimlock/framework/Slice.scala (hunk 3 above)
Comment (hgrano) [reply grimlock-core/src/main/scala/commbank/grimlock/framework/Slice.scala]: Ok I'll rename them

File: grimlock-core/src/main/scala/commbank/grimlock/framework/Slice.scala (hunk 1 above)
Comment (hgrano) [reply grimlock-core/src/main/scala/commbank/grimlock/framework/Slice.scala]: Having the implicit `ValueConstraints[P]` on the trait itself make the code a bit safer - in that we can never accidentally create an `implicit def` that resolves an instance of `IndexConstraints[P, _]` with `P` containing something other than `Value[_]`. But I will remove this `implicit val` in the interest of keeping the code clean

File: grimlock-core/src/main/scala/commbank/grimlock/framework/Slice.scala (hunk 3 above)
Comment (hgrano) [reply grimlock-core/src/main/scala/commbank/grimlock/framework/Slice.scala]: Done

File: grimlock-core/src/main/scala/commbank/grimlock/framework/Position.scala (hunk 6 above)
Comment (hgrano) [reply grimlock-core/src/main/scala/commbank/grimlock/framework/Position.scala]: Done

File: grimlock-core/src/main/scala/commbank/grimlock/framework/Position.scala (hunk 7 above)
Comment (hgrano) [reply grimlock-core/src/main/scala/commbank/grimlock/framework/Position.scala]: Done.

**Pull request merged**